src/sst/elements/ariel/arielrtlev.h

```diff
@@ -0,0 +1,165 @@
+// Copyright 2009-2020 Sandia Corporation. Under the terms
+// of Contract DE-NA0003525 with Sandia Corporation, the U.S.
+// Government retains certain rights in this software.
+//
+// Copyright (c) 2009-2020, Sandia Corporation
+// All rights reserved.
+//
+// Portions are copyright of other developers:
+// See the file CONTRIBUTORS.TXT in the top level directory
+// the distribution for more information.
+//
+// This file is part of the SST software package. For license
+// information, see the LICENSE file in the top level directory of the
+// distribution.
+
+
+#ifndef _H_SST_ARIEL_RTL_EVENT
+#define _H_SST_ARIEL_RTL_EVENT
+
+#include "arielevent.h"
+#include <unordered_map>
+#include <vector>
+#include <deque>
+#include <string>
+
+using namespace SST;
+
+namespace SST {
+namespace ArielComponent {
+
+typedef struct RtlSharedData {
+
+    void* rtl_inp_ptr;
+    void* rtl_ctrl_ptr;   
+    void* updated_rtl_params;
+
+    std::unordered_map<uint64_t, uint64_t>* pageTable; 
+    std::unordered_map<uint64_t, uint64_t>* translationCache; 
+    std::deque<uint64_t> *freePages;
```

Comment:
The problem with copying is that these STL data structures get self destroyed as soon as the sharedData is sent out as SST::Event, mostly because their scope ends (nature of STL data structures).  
Can we actually perform dynamic allocation of any shared data to be sent out as SST::Event? I do see some ptr_address in GpuCommand class of ariel_shmem.h. Another way could be copy these data to memH and send one (virtAddr, physAddr) pair and 'size' to osseous to fetch pageTable/freePage data structures and use that to do rest of the accesses. Like similar to sending just CR3 of paging structure. 
This won't be a problem in case of MMIO as the data will directly be moved to device memory.

Reply:
@shubhamn94 @gvoskuilen What was the decision on this?

Reply:
I am seeking confirmation for this, whether the above mentioned approach is fine.

Reply:
@shubhamn94 I'm trying to figure out why the scope ends so I may be missing something. Does the event not have it's own copy of the sharedData when you send? The issue is that the event is using a pointer to pass an STL container that does not belong to the event. If you do not serialize the container, the pointer will be invalid on a remote rank. If you do serialize the container, you'll have to delete the copy on a remote rank or you'll have a memory leak. But, a receiving component doesn't know whether an event has ever been serialized. So, you're left with always deleting, which might delete Ariel's page table if the event wasn't serialized. Or, never deleting, in which case you run the risk of never deleting copies if the event got serialized. I thought about adding a serializer function in SST Core for a pointer to a map/deque but it makes it really easy to create memory leaks like this.

An example of passing a copy of an STL container would be the std::vector payload_ field in memHierarchy/memEvent.h. Every new event has its own std::vector that is cleaned up when the event is deleted.

Writing the data structure into memory via memH and sharing it with the device would also be fine. It will potentially be a lot of writes if the page table is large, which could be slow.

GpuCommand is a little different as the pointer is an address in the simulated memory space to tell the Pin process where to find data in the simulated memory. It is not passing the data structure itself. You can pass pointers if you're careful that either the pointer alone (even on a remote rank) is all that's needed, not the data structure (like in GpuCommand or if you send the device a 'pointer' to the pageTable you've written into simulated memory) *or* that the data structure "belongs" to the event so that it's always safe to delete when the event is deleted.